### PR TITLE
Add sample rate and tags via new options parameter

### PR DESCRIPTION
**Pull request by fertel**
Sampling:
https://github.com/etsy/statsd/blob/master/docs/metric_types.md#sampling

Tagging:
http://docs.datadoghq.com/guides/dogstatsd/#fields-1

### CODE REVIEW

File: lib/statix.ex [hunk 1]
```diff
@@ -44,24 +44,28 @@ defmodule Statix do
...
-        |> Statix.transmit(:set, key, val)
+        |> Statix.transmit(:set, key, val, options)
       end
     end
   end
 
-  def transmit(conn, type, key, val) when is_binary(key) or is_list(key) do
-    Statix.Conn.transmit(conn, type, key, to_string(val))
+  def transmit(conn, type, key, val,options \\ []) when is_binary(key) or is_list(key) do
+    if options[:sample_rate] || 1 >  :rand.uniform do
```
Comment (lexmag) [inline lib/statix.ex]: Let's use `Keyword.get(options, :sample_rate, 1.0)` here.

File: lib/statix.ex [hunk 2]
```diff
@@ -44,24 +44,28 @@ defmodule Statix do
...
         @statix_conn
-        |> Statix.transmit(:set, key, val)
+        |> Statix.transmit(:set, key, val, options)
       end
     end
   end
 
-  def transmit(conn, type, key, val) when is_binary(key) or is_list(key) do
-    Statix.Conn.transmit(conn, type, key, to_string(val))
+  def transmit(conn, type, key, val,options \\ []) when is_binary(key) or is_list(key) do
```
Comment (lexmag) [inline lib/statix.ex]: It needs space between `,` and `options`.

File: lib/statix.ex [hunk 3]
```diff
@@ -44,24 +44,28 @@ defmodule Statix do
       for pipelining and easily wrapping existing code.
       """
       # TODO: Use `:erlang.monotonic_time/1` when we depend on Elixir ~> 1.2
-      def measure(key, fun) when is_function(fun, 0) do
+      def measure(key, fun, options \\ []) when is_function(fun, 0) do
```
Comment (lexmag) [inline lib/statix.ex]: It will more convenient if we make `options` second argument in `measure/3` function.

File: lib/statix/packet.ex [hunk 4]
```diff
@@ -15,8 +15,11 @@ defmodule Statix.Packet do
     ]
   end
 
-  def build(header, name, key, val) do
-    [header, key, ?:, val, ?| | metric_type(name)]
+  def build(header, name, key, val, options) do
+    [header, key, ?:, val,  ?|,  metric_type(name)]
+    |> set_option(:sample_rate,options[:sample_rate])
+    |> set_option(:tags,options[:tags])
+
```
Comment (lexmag) [inline lib/statix/packet.ex]: Unnecessary newline. ✂️

File: lib/statix/packet.ex [hunk 5]
```diff
@@ -29,4 +32,12 @@ defmodule Statix.Packet do
...
     defp metric_type(unquote(name)), do: unquote(type)
   end
+  defp set_option(stats, _, nil), do: stats
+  defp set_option(stats, :sample_rate,sample_rate) do
+    [stats| ["|@", :io_lib.format('~.2f', [sample_rate])]]
+  end
+  defp set_option(stats, :tags, tags) do
+    [stats| ["|#", Enum.join(tags,",") ]]
+  end
+
```
Comment (lexmag) [inline lib/statix/packet.ex]: Unnecessary newline. ✂️

File: lib/statix/packet.ex [hunk 6]
```diff
@@ -29,4 +32,12 @@ defmodule Statix.Packet do
   for {name, type} <- metrics do
     defp metric_type(unquote(name)), do: unquote(type)
   end
+  defp set_option(stats, _, nil), do: stats
+  defp set_option(stats, :sample_rate,sample_rate) do
+    [stats| ["|@", :io_lib.format('~.2f', [sample_rate])]]
```
Comment (lexmag) [inline lib/statix/packet.ex]: It needs space before` |`.

File: lib/statix/packet.ex [hunk 7]
```diff
@@ -29,4 +32,12 @@ defmodule Statix.Packet do
...
+  end
+  defp set_option(stats, :tags, tags) do
+    [stats| ["|#", Enum.join(tags,",") ]]
```
Comment (lexmag) [inline lib/statix/packet.ex]: It needs space before` |`.

Comment (lexmag) [inline lib/statix/packet.ex]: I think we need to escape `,` in tags, otherwise it will be interpreted incorrectly.

**Comment by fertel** [comment]
Made changes as per your suggestions. 

On another note - when load testing - granted I wasn't doing sampling. I have an experimental branch which solved some of the issues - Allowing you to batch a number of calls. Not sure if like the api for it.  I didn't put much thought into it - I wanted to get something up and running quick.

Essentially you can send a number of stats at once using - something like this:
```StatixModule.multi([{:counter, "foo.bar" 1.0}, {:histogram, "foo.baz", 2.0}])```

Let me know if you have any feedback. Would be happy to spend more time on it and put in another pull request if you thing it belongs.

You can see changes here:
https://github.com/adqio/statix/tree/multi

**Comment by lexmag** [comment]
Hey @fertel, I have just merged it! 👍

**Comment by lexmag** [comment]
Regarding "multi" support, I'm not sure yet how API should look like.
I think it requires some investigation on how other languages handle that, some overview will definitely help in defining next step.

**Pull request closed without merging**